src/Path/Scope.hs

```diff
@@ -1,47 +1,114 @@
 import Control.Monad ((>=>), guard)
 import Control.Monad.Module
 import Control.Monad.Trans (MonadTrans (..))
+import Data.Bifoldable
+import Data.Bifunctor
+import Data.Bitraversable
 import Data.Function (on)
 import Data.List (elemIndex)
 import GHC.Generics (Generic1)
+import Path.Pretty
 
-data Incr a b = Z a | S b
+data Var a b = B a | F b
   deriving (Eq, Foldable, Functor, Ord, Show, Traversable)
 
-instance Applicative (Incr a) where
-  pure = S
-  Z e <*> _ = Z e
-  S f <*> a = f <$> a
+instance (Pretty b, Pretty f) => Pretty (Var b f) where
+  pretty = \case
+    B b -> pretty b
+    F f -> pretty f
 
-instance Monad (Incr a) where
-  Z e >>= _ = Z e
-  S a >>= f = f a
+instance Bifoldable Var where
+  bifoldMap f g = \case
+    B a -> f a
+    F a -> g a
 
-match :: Applicative f => (b -> Either a c) -> b -> Incr a (f c)
-match f x = either Z (S . pure) (f x)
+instance Bitraversable Var where
+  bitraverse f g = \case
+    B a -> B <$> f a
+    F a -> F <$> g a
 
-matchM :: (Applicative f, Functor m) => (b -> m (Either a c)) -> b -> m (Incr a (f c))
-matchM f x = either Z (S . pure) <$> f x
+instance Bifunctor Var where
+  bimap f g = \case
+    B a -> B (f a)
+    F a -> F (g a)
+
+instance Applicative (Var a) where
+  pure = F
+  B e <*> _ = B e
+  F f <*> a = f <$> a
+
+instance Monad (Var a) where
+  B e >>= _ = B e
+  F a >>= f = f a
+
+var :: (a -> c) -> (b -> c) -> Var a b -> c
+var z s = \case { B a -> z a ; F b -> s b }
+
+match :: Applicative f => (b -> Either a c) -> b -> Var a (f c)
+match f x = either B (F . pure) (f x)
+
+matchM :: (Applicative f, Functor m) => (b -> m (Either a c)) -> b -> m (Var a (f c))
+matchM f x = either B (F . pure) <$> f x
 
 matchMaybe :: (b -> Maybe a) -> (b -> Either a b)
 matchMaybe f a = maybe (Right a) Left (f a)
 
-fromIncr :: (a -> b) -> Incr a b -> b
-fromIncr a = incr a id
 
-incr :: (a -> c) -> (b -> c) -> Incr a b -> c
-incr z s = \case { Z a -> z a ; S b -> s b }
+data Nat = Z | S Nat
+  deriving (Eq, Ord, Show)
+
+data Fin n where
+  FZ :: Fin ('S n)
+  FS :: Fin n -> Fin ('S n)
+
+deriving instance Eq   (Fin n)
+deriving instance Ord  (Fin n)
+deriving instance Show (Fin n)
+
+instance Pretty (Fin n) where
+  pretty = prettyVar . finToInt
+
+absurdFin :: Fin 'Z -> a
+absurdFin v = case v of {}
+
+finToInt :: Fin n -> Int
+finToInt FZ     = 0
+finToInt (FS n) = 1 + finToInt n
 
+strengthenFin :: Fin ('S n) -> Maybe (Fin n)
+strengthenFin FZ     = Nothing
+strengthenFin (FS n) = Just n
 
-newtype Scope a f b = Scope (f (Incr a (f b)))
+
+data Vec n a where
+  VZ :: Vec 'Z a
+  VS :: a -> Vec n a -> Vec ('S n) a
+
+deriving instance Eq   a => Eq   (Vec n a)
+deriving instance Ord  a => Ord  (Vec n a)
+deriving instance Show a => Show (Vec n a)
+
+deriving instance Foldable    (Vec n)
+deriving instance Functor     (Vec n)
+deriving instance Traversable (Vec n)
+
+(!) :: Vec n a -> Fin n -> a
+VS h _ ! FZ   = h
+VS _ t ! FS n = t ! n
+VZ     ! n    = absurdFin n
```

Comment:
I’m not actually certain why GHC can’t deduce this on its own; maybe because `Vec 'Z a` is inhabited and so even tho `Fin 'Z` isn’t, we might … do … something? Because laziness? I’m not sure.